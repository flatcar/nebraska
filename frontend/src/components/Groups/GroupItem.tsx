import CheckIcon from '@mui/icons-material/Check';
import CloseIcon from '@mui/icons-material/Close';
import ScheduleIcon from '@mui/icons-material/Schedule';
import { Box, Divider, Typography } from '@mui/material';
import Grid from '@mui/material/Grid';
import { TFunction } from 'i18next';
import React from 'react';
import { useTranslation } from 'react-i18next';
import _ from 'underscore';

import API from '../../api/API';
import { Group, VersionBreakdownEntry } from '../../api/apiDataTypes';
import { applicationsStore } from '../../stores/Stores';
import { useGroupVersionBreakdown } from '../../utils/helpers';
import ChannelItem from '../Channels/ChannelItem';
import { CardFeatureLabel, CardHeader, CardLabel } from '../common/Card';
import Empty from '../common/EmptyContent';
import ListItem from '../common/ListItem';
import MoreMenu from '../common/MoreMenu';
import VersionProgressBar from '../common/VersionBreakdownBar';

// From this number, we stop rate limiting in the backend
const MAX_UPDATES_PER_TIME_PERIOD = 900000;

export function formatUpdateLimits(t: TFunction, group: Group) {
  if (group.policy_max_updates_per_period >= MAX_UPDATES_PER_TIME_PERIOD) {
    return t('groups|update_policy_unlimited');
  }
  return t('groups|max_updates_per_period', {
    policy_max_updates_per_period: group.policy_max_updates_per_period,
    policy_period_interval: group.policy_period_interval,
  });
}

export interface GroupItemProps {
  group: Group;
  handleUpdateGroup: (appID: string, groupID: string) => void;
}

function GroupItem({ group, handleUpdateGroup }: GroupItemProps) {
  const { t } = useTranslation();
  const [totalInstances, setTotalInstances] = React.useState<null | number>(null);
  const versionBreakdown = useGroupVersionBreakdown(group);

  function deleteGroup(appID: string, groupID: string) {
    const confirmationText = t('groups|group_delete_confirmation');
    if (window.confirm(confirmationText)) {
      applicationsStore().deleteGroup(appID, groupID);
    }
  }

  React.useEffect(() => {
    API.getInstancesCount(group.application_id, group.id, '1d')
      .then(result => {
        setTotalInstances(result);
      })
      .catch(err => console.error('Error getting total instances in Group/Item', err));
  }, []);

  return (
    <PureGroupItem
      group={group}
      handleUpdateGroup={handleUpdateGroup}
      deleteGroup={deleteGroup}
      versionBreakdown={versionBreakdown}
      totalInstances={totalInstances}
    />
  );
}

export interface PureGroupItemProps {
  group: Group;
  versionBreakdown: VersionBreakdownEntry[] | null;
  totalInstances: number | null;
  handleUpdateGroup: (appID: string, groupID: string) => void;
  deleteGroup: (appID: string, groupID: string) => void;
}

export function PureGroupItem({
  group,
  versionBreakdown,
  totalInstances,
  handleUpdateGroup,
  deleteGroup,
}: PureGroupItemProps) {
  const { t } = useTranslation();

  const description = group.description || t('groups|description_none_provided');
  const channel = group.channel || null;

  const groupChannel = _.isEmpty(group.channel) ? (
    <CardLabel>{t('groups|channel_none_provided')}</CardLabel>
  ) : (
    <ChannelItem channel={group.channel} />
  );

  return (
    <ListItem
      disableGutters
      sx={{
        paddingBottom: 0,
      }}
    >
      <Grid container>
        <Grid size={12}>
          <CardHeader
            cardMainLinkLabel={group.name}
            cardMainLinkPath={`/apps/${group.application_id}/groups/${group.id}`}
            cardId={group.id}
            cardTrack={group.track}
            cardDescription={description}
          >
            <MoreMenu
              options={[
                {
                  label: t('frequent|edit'),
                  action: () => handleUpdateGroup(group.application_id, group.id),
                },
                {
                  label: t('frequent|delete'),
                  action: () => deleteGroup(group.application_id, group.id),
                },
              ]}
            />
          </CardHeader>
        </Grid>
        <Grid container justifyContent="space-between" size={12}>
          <Grid container direction="column" sx={{ padding: '1em' }} size={4}>
            <Grid>
              <CardFeatureLabel>{t('groups|instances')}</CardFeatureLabel>
              <Box>
                <CardLabel labelStyle={{ fontSize: '1.5rem' }}>
                  {totalInstances !== null ? (
                    totalInstances > 0 ? (
                      totalInstances
                    ) : (
                      t('frequent|none')
                    )
                  ) : (
                    <Empty>{t('frequent|loading')}</Empty>
                  )}
                </CardLabel>
                <Box display="flex" mr={2}>
                  <ScheduleIcon color="disabled" />
                  <Box pl={1} color="text.disabled">
                    <Typography
                      sx={{
                        color: 'rgba(0,0,0,0.6)',
                        fontSize: '0.875rem',
                        lineHeight: 2.0,
                      }}
                    >
                      {t('groups|time_last_24_hours')}
                    </Typography>
                  </Box>
                </Box>
              </Box>
            </Grid>
          </Grid>
          <Box width="1%">
            <Divider orientation="vertical" variant="fullWidth" />
          </Box>
          <Grid container direction="column" sx={{ padding: '1em' }} size={7}>
            <Grid>
              <CardFeatureLabel>{t('groups|channel')}</CardFeatureLabel> {groupChannel}
            </Grid>
            <Grid>
              <CardFeatureLabel>{t('groups|updates')}</CardFeatureLabel>
              <Box p={1} mb={1}>
                <CardLabel>
                  <Box display="flex">
                    {group.policy_updates_enabled ? (
                      <>
                        <Box>{t('frequent|enabled')}</Box>
                        <Box>
                          <CheckIcon
                            sx={{ color: theme => theme.palette.success.main }}
                            fontSize={'small'}
                          />
                        </Box>
                      </>
                    ) : (
                      <>
                        <Box>{t('frequent|disabled')}</Box>
                        <Box>
                          <CloseIcon color="error" />
                        </Box>
                      </>
                    )}
                  </Box>
                </CardLabel>
              </Box>
            </Grid>
            <Grid>
              <CardFeatureLabel>{t('groups|rollout_policy')}</CardFeatureLabel>
              <Box p={1} mb={1}>
                <CardLabel>{formatUpdateLimits(t, group)}</CardLabel>
              </Box>
            </Grid>
            <Grid container>
              <Grid size={12}>
                <CardFeatureLabel>{t('groups|version_breakdown_lower')}</CardFeatureLabel>
              </Grid>
              <Grid size={12}>
                {versionBreakdown === null ? (
                  <Empty>{t('frequent|loading')}</Empty>
                ) : versionBreakdown?.length > 0 ? (
                  <VersionProgressBar version_breakdown={versionBreakdown} channel={channel} />
                ) : (
                  <Empty>{t('groups|instances_none_available')}</Empty>
                )}
              </Grid>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </ListItem>
  );
}

export default GroupItem;
