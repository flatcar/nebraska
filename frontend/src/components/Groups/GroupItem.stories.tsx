import { Meta, StoryFn } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';

import { PureGroupItem, PureGroupItemProps } from './GroupItem';

export default {
  title: 'groups/GroupItem',
  argTypes: {
    handleUpdateGroup: { action: 'handleUpdateGroup' },
    deleteGroup: { action: 'deleteGroup' },
  },
} as Meta;

const Template: StoryFn<PureGroupItemProps> = args => {
  return (
    <MemoryRouter>
      <PureGroupItem {...args} />
    </MemoryRouter>
  );
};

export const Group = {
  render: Template,

  args: {
    versionBreakdown: [],
    totalInstances: 2,
    group: {
      id: '11a585f6-9418-4df0-8863-78b2fd3240f8',
      name: 'Stable (ARM)',
      description: 'For production clusters (ARM)',
      created_ts: '2015-09-19T07:09:34.269062+02:00',
      rollout_in_progress: false,
      application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
      channel_id: '5dfe7b12-c94a-470d-a2b6-2eae78c5c9f5',
      policy_updates_enabled: true,
      policy_safe_mode: false,
      policy_office_hours: false,
      policy_timezone: 'Europe/Berlin',
      policy_period_interval: '1 minutes',
      policy_max_updates_per_period: 999999,
      policy_update_timeout: '60 minutes',
      channel: {
        id: '5dfe7b12-c94a-470d-a2b6-2eae78c5c9f5',
        name: 'stable',
        color: '#1458d6',
        created_ts: '2015-09-19T07:09:34.261241+02:00',
        application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
        package_id: null,
        package: null,
        arch: 2,
      },
      track: 'stable',
    },
  },
};

export const Loading = {
  render: Template,

  args: {
    versionBreakdown: null,
    totalInstances: null,
    group: {
      id: '11a585f6-9418-4df0-8863-78b2fd3240f8',
      name: 'Stable (ARM)',
      description: 'For production clusters (ARM)',
      created_ts: '2015-09-19T07:09:34.269062+02:00',
      rollout_in_progress: false,
      application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
      channel_id: '5dfe7b12-c94a-470d-a2b6-2eae78c5c9f5',
      policy_updates_enabled: true,
      policy_safe_mode: false,
      policy_office_hours: false,
      policy_timezone: 'Europe/Berlin',
      policy_period_interval: '1 minutes',
      policy_max_updates_per_period: 999999,
      policy_update_timeout: '60 minutes',
      channel: {
        id: '5dfe7b12-c94a-470d-a2b6-2eae78c5c9f5',
        name: 'stable',
        color: '#1458d6',
        created_ts: '2015-09-19T07:09:34.261241+02:00',
        application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
        package_id: null,
        package: null,
        arch: 2,
      },
      track: 'stable',
    },
  },
};
