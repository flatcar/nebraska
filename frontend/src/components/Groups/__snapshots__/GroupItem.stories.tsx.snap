// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groups/GroupItem Group smoke-test 1`] = `
<li class="MuiListItem-root MuiListItem-padding MuiListItem-divider stable-makeStyles-divider-5 stable-makeStyles-root-1 css-1oz7m58-MuiListItem-root"
    data-testid="list-item"
>
  <div class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root">
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-gridHeader-6 css-1xzqqmo-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-j1w48a-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <h2 class="MuiTypography-root MuiTypography-h2 css-150084s-MuiTypography-root">
              <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-mainLink-8 css-1emptwe-MuiTypography-root-MuiLink-root"
                 href="/apps/e96281a6-d1af-4bde-9a0a-97b76e56dc57/groups/11a585f6-9418-4df0-8863-78b2fd3240f8"
              >
                Stable (ARM)
              </a>
            </h2>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-1yq8yj7">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap stable-makeStyles-idLabel-11 css-1s94icn-MuiTypography-root"
                 arial-label="group-id"
              >
                stable
              </p>
            </div>
            <div class="MuiBox-root css-1yuhvjn">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-descriptionLabel-10 css-1s23nr9-MuiTypography-root">
                For production clusters (ARM)
              </span>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <div>
            <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1pux6rn-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                    aria-controls="simple-menu-1"
                    aria-haspopup="true"
                    aria-label="Open menu"
                    data-testid="more-menu-open-button"
            >
              <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="MoreVertIcon"
              >
                <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
                </path>
              </svg>
              <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
              </span>
            </button>
          </div>
        </div>
      </div>
      <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light stable-makeStyles-innerDivider-12 css-xetjfm-MuiDivider-root">
    </div>
    <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-18n0oo3-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-4 stable-makeStyles-itemSection-2 css-13q96zz-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-9 css-1s23nr9-MuiTypography-root">
            Instances
          </span>
          <div class="MuiBox-root css-0">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-13 stable-makeStyles-label-17 css-1s23nr9-MuiTypography-root">
              2
            </span>
            <div class="MuiBox-root css-f0oijv">
              <svg class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="ScheduleIcon"
              >
                <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                </path>
                <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                </path>
              </svg>
              <div class="MuiBox-root css-1hsc5oo">
                <p class="MuiTypography-root MuiTypography-body1 stable-makeStyles-last24hours-4 css-1s23nr9-MuiTypography-root">
                  last 24 hours
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="MuiBox-root css-1pelcm7">
        <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-ftc7wc-MuiDivider-root">
      </div>
      <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-7 stable-makeStyles-itemSection-2 css-n5qnn3-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-9 css-1s23nr9-MuiTypography-root">
            Channel
          </span>
          <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1i06p90-MuiListItem-root">
            <div class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root">
              <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault stable-makeStyles-colorAvatar-20 stable-makeStyles-colorAvatar-21 css-1onlcy6-MuiAvatar-root">
                  s
                </div>
              </div>
              <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                <div class="MuiListItemText-root MuiListItemText-multiline stable-makeStyles-root-19 css-konndc-MuiListItemText-root">
                  <div class="MuiBox-root css-70qvj9">
                    <div class="MuiBox-root css-1xsd608">
                      stable
                    </div>
                  </div>
                  <div class="MuiBox-root css-15aqakd">
                    <div class="MuiBox-root css-0">
                      No package (ARM64)
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-9 css-1s23nr9-MuiTypography-root">
            Updates
          </span>
          <div class="MuiBox-root css-1hply4a">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-13 stable-makeStyles-label-23 css-1s23nr9-MuiTypography-root">
              <div class="MuiBox-root css-k008qs">
                <div class="MuiBox-root css-0">
                  Enabled
                </div>
                <div class="MuiBox-root css-0">
                  <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall stable-makeStyles-success-3 css-ptiqhd-MuiSvgIcon-root"
                       focusable="false"
                       aria-hidden="true"
                       viewbox="0 0 24 24"
                       data-testid="CheckIcon"
                  >
                    <path d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z">
                    </path>
                  </svg>
                </div>
              </div>
            </span>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-9 css-1s23nr9-MuiTypography-root">
            Rollout Policy
          </span>
          <div class="MuiBox-root css-1hply4a">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-13 stable-makeStyles-label-25 css-1s23nr9-MuiTypography-root">
              Unlimited number of parallel updates
            </span>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-9 css-1s23nr9-MuiTypography-root">
              Version breakdown
            </span>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-19midj6">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-ii1goy-MuiTypography-root"
                 data-testid="empty"
              >
                No instances available.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;

exports[`groups/GroupItem Loading smoke-test 1`] = `
<li class="MuiListItem-root MuiListItem-padding MuiListItem-divider stable-makeStyles-divider-31 stable-makeStyles-root-27 css-1oz7m58-MuiListItem-root"
    data-testid="list-item"
>
  <div class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root">
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-gridHeader-32 css-1xzqqmo-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-j1w48a-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <h2 class="MuiTypography-root MuiTypography-h2 css-150084s-MuiTypography-root">
              <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-mainLink-34 css-1emptwe-MuiTypography-root-MuiLink-root"
                 href="/apps/e96281a6-d1af-4bde-9a0a-97b76e56dc57/groups/11a585f6-9418-4df0-8863-78b2fd3240f8"
              >
                Stable (ARM)
              </a>
            </h2>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-1yq8yj7">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap stable-makeStyles-idLabel-37 css-1s94icn-MuiTypography-root"
                 arial-label="group-id"
              >
                stable
              </p>
            </div>
            <div class="MuiBox-root css-1yuhvjn">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-descriptionLabel-36 css-1s23nr9-MuiTypography-root">
                For production clusters (ARM)
              </span>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <div>
            <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1pux6rn-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                    aria-controls="simple-menu-2"
                    aria-haspopup="true"
                    aria-label="Open menu"
                    data-testid="more-menu-open-button"
            >
              <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="MoreVertIcon"
              >
                <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
                </path>
              </svg>
              <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
              </span>
            </button>
          </div>
        </div>
      </div>
      <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light stable-makeStyles-innerDivider-38 css-xetjfm-MuiDivider-root">
    </div>
    <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-18n0oo3-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-4 stable-makeStyles-itemSection-28 css-13q96zz-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-35 css-1s23nr9-MuiTypography-root">
            Instances
          </span>
          <div class="MuiBox-root css-0">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-39 stable-makeStyles-label-43 css-1s23nr9-MuiTypography-root">
              <div class="MuiBox-root css-19midj6">
                <p class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-ii1goy-MuiTypography-root"
                   data-testid="empty"
                >
                  Loading…
                </p>
              </div>
            </span>
            <div class="MuiBox-root css-f0oijv">
              <svg class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="ScheduleIcon"
              >
                <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                </path>
                <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                </path>
              </svg>
              <div class="MuiBox-root css-1hsc5oo">
                <p class="MuiTypography-root MuiTypography-body1 stable-makeStyles-last24hours-30 css-1s23nr9-MuiTypography-root">
                  last 24 hours
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="MuiBox-root css-1pelcm7">
        <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-ftc7wc-MuiDivider-root">
      </div>
      <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-7 stable-makeStyles-itemSection-28 css-n5qnn3-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-35 css-1s23nr9-MuiTypography-root">
            Channel
          </span>
          <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1i06p90-MuiListItem-root">
            <div class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root">
              <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault stable-makeStyles-colorAvatar-46 stable-makeStyles-colorAvatar-47 css-1onlcy6-MuiAvatar-root">
                  s
                </div>
              </div>
              <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                <div class="MuiListItemText-root MuiListItemText-multiline stable-makeStyles-root-45 css-konndc-MuiListItemText-root">
                  <div class="MuiBox-root css-70qvj9">
                    <div class="MuiBox-root css-1xsd608">
                      stable
                    </div>
                  </div>
                  <div class="MuiBox-root css-15aqakd">
                    <div class="MuiBox-root css-0">
                      No package (ARM64)
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-35 css-1s23nr9-MuiTypography-root">
            Updates
          </span>
          <div class="MuiBox-root css-1hply4a">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-39 stable-makeStyles-label-49 css-1s23nr9-MuiTypography-root">
              <div class="MuiBox-root css-k008qs">
                <div class="MuiBox-root css-0">
                  Enabled
                </div>
                <div class="MuiBox-root css-0">
                  <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall stable-makeStyles-success-29 css-ptiqhd-MuiSvgIcon-root"
                       focusable="false"
                       aria-hidden="true"
                       viewbox="0 0 24 24"
                       data-testid="CheckIcon"
                  >
                    <path d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z">
                    </path>
                  </svg>
                </div>
              </div>
            </span>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-35 css-1s23nr9-MuiTypography-root">
            Rollout Policy
          </span>
          <div class="MuiBox-root css-1hply4a">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-39 stable-makeStyles-label-51 css-1s23nr9-MuiTypography-root">
              Unlimited number of parallel updates
            </span>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-35 css-1s23nr9-MuiTypography-root">
              Version breakdown
            </span>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-19midj6">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-ii1goy-MuiTypography-root"
                 data-testid="empty"
              >
                Loading…
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;
